### PR TITLE
Maintenance tasks for repository lmms

### PR DESCRIPTION
**Pull request by jasp00**
This adds an hourly cron job to maintain the repository `lmms`. Is an access to GitHub set up already? A commit preparing the affected files should be done first.

### CODE REVIEW

**Comment by jasp00** [comment]
>Hourly may be overkill, don't you think?

It is only a `git pull` and the scripts are quick. I could avoid the scripts though, if there were no errors.

**Comment by jasp00** [comment]
>I could avoid the scripts though, if there were no errors.

Done.

**Comment by tresf** [comment]
* Current `cron` file naming convention looks like this currently: `update_community_feeds`.  Please replace dashes with underscores.
 * The file name `lmms-tasks` should be more descriptive, such as `git_maintenance_tasks`
 * Although the intent of `maintenance-tasks-done` seems to be to make sure it runs at least once, the code suggests it's not needed; it'll hit `else` on the `if [ -e lmms ]` block.
 * To be safe, `git clone [...]` should explicitly specify a branch.   I recommend master.
 * `Already up-to-date.` should set terminal to English  to make sure this parses the English output (e.g. `LC_ALL=C`)

I like the thought that went into the `cron/tasks` structure and I don't mean to be overly critical, but I feel this could all be in one script.

For example... here's the entire task in 34 lines with comments.   I haven't tested the below code but when run interactively, it should be apparent what it's doing for support purposes.

```bash
#!/bin/sh

# Exit on any non-zero return codes; force en_US locale
set -e && LC_ALL=C
mkdir -p ~/lmms-repo && cd ~/lmms-repo

if [ -e lmms ]; then
	cd lmms && PULLRES=$(git pull)
	if [ "$PULLRES" = "Already up-to-date." ]; then
		exit
	fi
else
	git clone -b master https://github.com/LMMS/lmms && cd lmms
fi

# Update CONTRIBUTORS
git shortlog -sne | cut -c 8- > doc/CONTRIBUTORS

# Update PROJECT_YEAR 
YEAR=$(date -u +%Y)
while true; do
	if [ "$(git log -n 1 --since=$YEAR-01-01T00:00:00Z)" ]; then
		break
	fi
	YEAR=$((YEAR - 1))
done

sed -e "/^SET(PROJECT_YEAR / c \SET(PROJECT_YEAR $YEAR)" -i CMakeLists.txt

# Commit changes, if any
if [ "$(git diff)" ]; then
	git add -u && git commit -m "Maintenance tasks"
	git push origin master
fi
```

**Comment by tresf** [comment]
>  Is an access to GitHub set up already?

No, not yet.  `lmmsservice` is our GitHub service account.  I'd be happy to set it up with a pointer in the right direction.  Alternately, we can add your public key to the server if you prefer to test it all out anyway.

**Comment by jasp00** [comment]
>* Although the intent of `maintenance-tasks-done` seems to be to make sure it runs at least once, the code suggests it's not needed

It is needed to run the tasks at most once, otherwise the tasks would run each hour.

>I feel this could all be in one script.

The point is to easily add, enable and debug tasks; they are like standalone plug-ins. Each task should have its own commit actually. More tasks are expected to be developed.

>we can add your public key to the server

This is an option, but not necessary.

>I'd be happy to set it up with a pointer in the right direction.

Option 1, with SSH:

1. Generate 4096-bit RSA SSH keys.
2. Upload the public key to GitHub.
3. Configure `~/.ssh/config`.

Option 2, with password:

1. Configure `~/.netrc`.

**Comment by tresf** [comment]
> It is needed to run the tasks at most once, otherwise the tasks would run each hour.

If it only commits changes, I don't see how this statement is valid.

> The point is to easily add, enable and debug tasks; they are like standalone plug-ins. Each task should have its own commit actually. 

That's an opinion.  As the site maintainer, I won't merge it in this condition.  If another maintainer disagrees you can work with them.

> More tasks are expected to be developed.

This is speculative.

> Option 1, with SSH:

Thanks.  I'll post back to this thread once setup.

**Comment by jasp00** [comment]
The goal is scalability.

>>otherwise the tasks would run each hour.

>If it only commits changes, I don't see how this statement is valid.

The tasks would not make any commit, but they would still run (do you [remember](https://github.com/LMMS/lmms.io/pull/198#issuecomment-248616034)?).

>>The point is to easily add, enable and debug tasks;

>That's an opinion.

Why would you not want a system that is based on `run-parts`, which is easy to debug single tasks, to disable with a file rename, to expand with independent executables, etc.? Why would you want a system that is not scalable?

>>More tasks are expected to be developed.

>This is speculative.

I [mentioned](https://github.com/LMMS/lmms/pull/3016#issuecomment-274929258) some possible tasks: remove executable bits, optimize PNG images, format code, etc. Do you want me to add another task?

**Comment by tresf** [comment]
> The tasks would not make any commit, but they would still run (do you remember?).

What you're referencing is the question if hourly was overkill.   Your argument I felt was valid.

> jasp00 wrote:  "It is only a git pull and the scripts are quick. I could avoid the scripts though, if there were no errors."

Still, I don't like magic lock files. They're bad design as they require manual cleanup.  Last, I don't even understand the purpose of this one.  I can't see a logical scenario when it's warranted.

> I mentioned some possible tasks: remove executable bits, optimize PNG images, format code, etc. Do you want me to add another task?

Very good points.  Sorry for the delay on the SSH push config.  Will set that up ASAP.

**Comment by jasp00** [comment]
>Still, I don't like magic lock files. They're bad design as they require manual cleanup when things fail.

There is no cleanup, lock files do not need to be removed. These locks are released when the process finishes. AFAIK, this is the best way to have only one instance per service.

> Last, I don't even understand the purpose of this one. I can't see a logical scenario when it's warranted.

It is defensive coding. The case is when the process lasts more than an hour somehow. This may happen if networking goes slow, if the server is heavy loaded, if a task is deadlocked, if we include 60 one-minute tasks, etc. If this happens, we do not want a second instance to interfere with the first one.

I should add a protection against merge conflicts.

**Comment by jasp00** [comment]
>I should add a protection against merge conflicts.

In the future, it is very unlikely right now.

**Comment by tresf** [comment]
> I should add a protection against merge conflicts.

> > In the future, it is very unlikely right now.

As is a 1-hour-long merge.

**Comment by jasp00** [comment]
>As is a 1-hour-long merge.

1-hour-long process actually; I know it is unlikely. I would implement a protection against merge conflicts, but I have more pressing issues right now. The point is I would not reject an improvement.

**Comment by tresf** [comment]
@jasp00 SSH authentication is complete.

 * Pushes will authenticate automatically;  If for some reason they don't, we can prepend `ssh-agent` to the `git push` command.
 * Pushes must go to ssh instead of the default HTTP push.
   ```bash
   # example
   git remote set-url origin git@github.com:lmms/lmms.git
   ```
 * Here is a preview of what the push will look like: https://github.com/tresf/lmms/commit/c7f025f17d45f84215b93dcfd24eaa7e9ada7f2f

**Comment by jasp00** [comment]
We should test with a branch first, then switch to master, do you agree?

**Comment by tresf** [comment]
> We should test with a branch first, then switch to master, do you agree?

Sure.

**Comment by jasp00** [comment]
Ready.

**Comment by tresf** [comment]
Should kick off in 28 minutes... 

@jasp00 why are the scripts in `/dev` and `/tasks` and not `/cron` ?

**Comment by tresf** [comment]
Answering my own question, dev was already there.  I'm still confused as to the need for the new root directory.  Seem like `/dev` could have been the home for tasks.

**Comment by tresf** [comment]
Ok... @Lukas-W put the other script in `/dev` for a different purpose -- to help him test locally.  We need to move stuff around.  `/cron` can house everything and `/dev` should go back to just developer helpers/tools.

**Comment by jasp00** [comment]
>I'm still confused as to the need for the new root directory `/tasks`.

I chose `/tasks` to be independent of the trigger, either cron job or webhook.

>We need to move stuff around.

Sure.

There are [no commits](https://github.com/LMMS/lmms/commits/maintenance-test) in the test branch. Are you sure cron is running? Should I prepend `ssh-agent`?

**Comment by tresf** [comment]
Cron is running, there was just an authentication problem which has been fixed https://github.com/LMMS/lmms.io/commit/d2d67040716384a3a9ecee8fc962a37e700ffd89.

Note, the task was run interactively for testing purposes; the commits (https://github.com/LMMS/lmms/commit/b990b182962c74ea7278ebfb94bc1d02086b3f67, https://github.com/LMMS/lmms/commit/bf0e650ae8b75bcb562aff8d9d754e011b5f06a1),  were done with the interactive shell and on first clone but initial tests pass.

File: dev/git_maintenance_tasks [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# Force C locale
+export LC_ALL=C
+
+mkdir -p ~/lmms-repo
+cd ~/lmms-repo
+
+if [ -e lmms ]
+then
+	cd lmms
+	PULLRES=$(git pull)
```
Comment (tresf) [inline dev/git_maintenance_tasks]: @jasp00 should we consider appending `git checkout $BRANCH` to force us back onto `master` once ready?

Comment (jasp00) [inline dev/git_maintenance_tasks]: Sure, #214.

**Comment by tresf** [comment]
I can't find your most recent comments about the job failing but interactively tests pass, then when I scheduled the hourly to run now, it reverted the branch to the past somehow.  https://github.com/LMMS/lmms/blob/maintenance-test/doc/CONTRIBUTORS

**Comment by tresf** [comment]
Ok... @jasp00 I think we've hit a nuance with `git add`.

http://stackoverflow.com/questions/348170/how-to-undo-git-add-before-commit

I can reproduce the flapping behavior of `HEAD` and reset the file.  We can either try moving `git add` inside the conditional block, or try `git reset <file>` to remove it although there's a [very highly up-voted answer](http://stackoverflow.com/a/682343/3196753) that instead says we should use `git rm --cached <file>`.

**Comment by jasp00** [comment]
>Ok... @jasp00 I think we've hit a nuance with `git add`.

I do not think so, `git add` works as intended. See https://github.com/LMMS/lmms.io/pull/215#issuecomment-280304980.

**Pull request merged**